collect_app/src/main/java/org/odk/collect/android/map/MapboxMapFragment.java

```diff
@@ -175,6 +215,78 @@
         return new Style.Builder().fromUrl(url == null ? DEFAULT_STYLE_URL : url);
     }
 
+    @SuppressWarnings("TimberExceptionLogging")
+    private void addMbtiles(Style style, String id, File file) {
+        MbtilesFile mbtiles;
+        try {
+            mbtiles = new MbtilesFile(file);
+        } catch (MbtilesFile.UnsupportedFormatException e) {
+            Timber.w(e.getMessage());
+            return;
+        }
+
+        TileSet tileSet = createTileSet(mbtiles, tileServer.getUrlTemplate(id));
+        tileServer.addSource(id, mbtiles);
+
+        if (mbtiles.getType() == MbtilesFile.Type.VECTOR) {
+            style.addSource(new VectorSource(id, tileSet));
+            List<MbtilesFile.VectorLayer> layers = mbtiles.getVectorLayers();
+            for (MbtilesFile.VectorLayer layer : layers) {
+                // Pick a colour that's a function of the filename and layer name.
+                int hue = (((id + "." + layer.name).hashCode()) & 0x7fffffff) % 360;
```

Comment:
This may look a little strange.

We don't have a way for the user to configure which colours to use for drawing which layers; that would be a substantial UX effort that we aren't taking on yet.  So, the goal here is just to pick different colours for various layers, in a way that is deterministic for a given file and layer.  Using the hash code means that colours will appear effectively random to the user.

Reply:
Thanks for that extra explanation, that's a nice approach.